Introduce Domain tests using behat

matks opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Introduce new domain tests structure, setup, scripts
| Type?         | refacto
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10733
| How to test?  | Travis tests should be green and include the new unit test

⚠️ This needs to add a new Travis build.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/11560)
<!-- Reviewable:end -->

Commit:
TE: change position, unhhok and transplant a module

Commit:
TE: Add, edit, delete and bulk actions zone

CaptainYouz committed:
ui(FO | Product): remove image margin on product card view. #9738

CaptainYouz committed:
ui(FO | Home): center thumbnail image in case of small ratio. #9738

tomas862 committed:
creates default currency  controller

tomas862 committed:
displays currency list with basic data

tomas862 committed:
adds shop context restriction

tomas862 committed:
implements sorting functionality for exchange rates

tomas862 committed:
implements filtering by status

tomas862 committed:
adds currency toggling command ands its implementation

tomas862 committed:
adds extension to enable status toggling

tomas862 committed:
creates command for currency deletion

tomas862 committed:
adds constant for cannot disable default currency adds some todos

tomas862 committed:
implements default currency validation in toggle status handler

tomas862 committed:
implements delete action using command, implements default currency validation

tomas862 committed:
improves some sql code for shop context

tomas862 committed:
improves sql regarding the shop context

Commit:
adds create action, adds links which points to legacy form, adds security checks, adds demo restricted checks

Commit:
uses decorator to override currency list data provider

Commit:
moves currency grid data factory to core namespace

Commit:
overrides records to display data identical on the legacy currency pag

Commit:
implements search by iso code

Commit:
fixes some design issues, uses COUNT instead of FETCH_ROWS

Commit:
prevents translations in the options

tomas862 committed:
removes constructor tags

tomas862 committed:
adds new exceptions , removes exception constants

tomas862 committed:
renamed some tempaltes, changed names for some classes

tomas862 committed:
moves add new currency button definition to template

tomas862 committed:
changed some route path names

tomas862 committed:
adds missing route methods, changes main route name

tomas862 committed:
fixes some critical errors

tomas862 committed:
adds redirectRoute to annotations

tomas862 committed:
adds yesnochoice type

tomas862 committed:
adds some comments

tomas862 committed:
adds link column usage

Commit:
applied php-cs-fixer

sarjon committed:
add category cover image deleting command

sarjon committed:
add category menu thumbnail deleting

sarjon committed:
implement image deleting actions

sarjon committed:
handle image deleting exceptions

sarjon committed:
minor fixes & access control checks

sarjon committed:
redirect user to category after add/edit actions

sarjon committed:
cs fixer

sarjon committed:
improve menu thumbnail uploading

sarjon committed:
build assets

sarjon committed:
implement add customer command

sarjon committed:
add CustomerType for forms

sarjon committed:
implement edit customer command

sarjon committed:
add getting customer for editing

sarjon committed:
implement form rendering with default data

sarjon committed:
implement b2b part of customer form

sarjon committed:
add b2b customer suppor type

sarjon committed:
make password field requirement configurable for customer

matks commented:
@Quetzacoalt91 I prepared this PR to work with a new build on Travis, but this is just a suggestion, it can change if you think it's not a good idea.

sarjon committed:
render customer form with correct ui

Quetzacoalt91 commented:
If you think that is justified, go for it. This will obviously require more time to the whole process to complete (env preparation), but I understood you wanted to reduce it by optimizing the existing tests. :)
Keep in mind you have to update the `.travis.yml` file to add the `integration` step to the matrix.

sarjon committed:
make customer birthday field single widget

sarjon committed:
add VOs for customer email, firstname, lastname and password

sarjon committed:
improve category id validation

sarjon committed:
fix thumbnail id validation

sarjon committed:
add b2b customer fields to edit customer command

sarjon committed:
implement edit customer command dispatching when customer edit form is submitted

matks commented:
@Quetzacoalt91 The idea is to replace existing tests by new tests with a better run strategy (this will provide better performances) but this will happen in a long time 😅 

Maybe it's better for me to make these Integration tests run after the unit tests then ? We keep then the split "unit test" = "now browser" and "e2e test" = "with a browser"

Fabuloops committed:
Fix jquery shortcut function 'load' deprecate

mickaelandrieu commented:
ping @eternoendless, I think the pull request is right but this is not what we need to validate here.

Do you want to introduce Behat in 1.7.x or not? I think this is the right moment to decide :+1:

sarjon committed:
implement b2b fields editing using command

sarjon committed:
add b2b customer data to form provider

sarjon committed:
update AddCustomerCommand handler

sarjon committed:
add assertion for customer email when editing

sarjon committed:
add error handling in customer controller

sarjon committed:
allow creating customer when shop is selected

sarjon committed:
update edit customer command building

sarjon committed:
use vo's instead of plain values when creating customer

sarjon committed:
mark required fields with asterisks

sarjon committed:
add password hashing on customer add/edit

matks commented:
https://github.com/PrestaShop/PrestaShop/pull/11560/commits/16977c0ca3afe919b00e8af322f0c6b237c896e8 introduces a Command test and the required behat hooks to reset the database as the Command modifies the database.

2 database reset hooks are available:
- `@database-light-reset` : scenarios tagged like this will perform their mysql into a `START TRANSACTION ... ROLLBACK` statement so database is not modified during the test
- `@database-hard-reset` : scenarios tagged like this will see the database reset at the end of the scenario using the time-consuming `DatabaseCreator::restoreTestDB()` function

So the light reset is faster, however it comes with a known limitation:

Since Command Handlers use ObjectModels, it's the legacy core `\Db` class that is being used so the connection used to start/rollback the SQL transaction is the one from `\Db`

**This means all queries must be performed through this connection.**

Example of a failing usecase:
1) scenario starts: we execute `START TRANSACTION` on PDO connection from `\Db` class
2) data are inserted into the database through the PDO connection of `\Db`
3) a COUNT query is performed using PDO connection of Doctrine in Symfony code
4) scenario ends: `ROLLBACK` performed through `\Db` connection

Here, step 3) result is wrong because. To be right, it must be performed through the PDO connection of  `Db` too.

sarjon committed:
add some unit tests

sarjon committed:
add more unit tests

sarjon committed:
add more unit tests

sarjon committed:
do not diplsy password input when editing guest customer

sarjon committed:
add group selection in request when handling form submit

sarjon committed:
improve exception translation to user friendly errors

sarjon committed:
compare current customer email with new one when editing customer

sarjon committed:
fix birthday saving on add/edit for customer

sarjon committed:
display minimum required password length

sarjon committed:
enable ChoiceTable for add/edit forms

matks commented:
As discussed with @eternoendless and @PierreRambaud yesterday, we will avoid using traits and use abstract contexts instead

tomlev committed:
fix wrong discount line calculation on order/invoice

sarjon committed:
improve validation of customers birthday

sarjon committed:
keep allowed first and last names chars in variable

sarjon committed:
cs fixer

tomlev committed:
fix smarty in context mocker

tomlev committed:
test order to cart when processing checkout

tomlev committed:
oops

tomlev committed:
cs-fixer

tomlev committed:
fix php < 7

sarjon committed:
handle duplicate email exception

sarjon committed:
handle group access selection exception

sarjon committed:
add ape code validation

sarjon committed:
remove margin from material choice table

sarjon committed:
cs fixer

sarjon committed:
minor fixes

sarjon committed:
cs fixer

tomlev committed:
fix big amount voucher

tomlev committed:
fix ternary operator

sarjon committed:
fix password editing for customer

sarjon committed:
redirect user to customers list after successful edit

YosraAk committed:
TE: add delivery slip test

sarjon committed:
make category id validation more strict

tomlev committed:
remove init mail send from test

Commit:
TE: Credit slip

sarjon committed:
use legacy layout for birthday input

sarjon committed:
upate form rending with macro helper

sarjon committed:
add new error messages

Klemart3D committed:
Add customer_message table name into SQL request

To prevent "Column 'message' in field list is ambiguous"
See issue: #11849

MathiasReker committed:
Replace tabs with space in twig files

MathiasReker committed:
Replace tabs with spaces

MathiasReker committed:
php unit set up tear down visibility

MathiasReker committed:
Order imports

MathiasReker committed:
Remove unneeded curly braces

MathiasReker committed:
Remove unneeded control parentheses

MathiasReker committed:
fix mistake

MathiasReker committed:
Remove comments

Quetzacoalt91 committed:
Merge pull request #11745 from web-fu/hotfix/issue-11743-Wrong-Recovery-Password-Error-Messages

Fix for password recovery error messages

Quetzacoalt91 committed:
Merge pull request #11851 from PierreRambaud/fix/11801

An error is displayed "getimagesize(): php_network_getaddresses" when we try to change the status of the order to payment accepted

Quetzacoalt91 committed:
Merge pull request #11661 from jolelievre/reset-quantity

Reset quantity value to min value after adding to cart

Quetzacoalt91 committed:
Merge pull request #11857 from MathiasReker/t5

Replace tabs with spaces

Quetzacoalt91 committed:
Merge pull request #11869 from MathiasReker/t16

Remove unneeded control parentheses

tomas862 committed:
adds reset and redirect routes via route parameters and not injecting it via the constructor

Quetzacoalt91 committed:
Merge pull request #11867 from MathiasReker/t15

Remove unneeded curly braces

tomas862 committed:
fixes currencies list display

tomas862 committed:
fixes javascript functionality

Dan-Orchard committed:
CO: Fix lang table multistore data issue

tomas862 committed:
decreases input size to 120px

tomas862 committed:
adds placeholder for iso text input

sarjon committed:
add additional constraints for customer form

sarjon committed:
redirect to list after customer creation

Commit:
Do not allow null for low_stock_alert

Dan-Orchard committed:
moved id_shop_list

Quetzacoalt91 committed:
Merge pull request #11866 from MathiasReker/t14

Order imports

Quetzacoalt91 committed:
Merge pull request #11884 from eternoendless/fix-nullable

Do not allow null for low_stock_alert

Commit:
TE: Generate invoice pdf by date and status

sarjon committed:
fix missing const definition

sarjon committed:
fix categories grid id

mbadrani committed:
Merge pull request #11891 from hadrich-hatem/creditSlip

Credit slip, credit slip pdf and credit slip options

matks committed:
Fix travis script for running tests

mbadrani committed:
Merge pull request #11897 from hadrich-hatem/invoice_pdf

Generate invoice pdf by date and status

PierreRambaud committed:
Merge pull request #11895 from sarjon/fix-categories-grid-id

Fix categories grid id

matks committed:
Enable HelperList title_icon property

Quetzacoalt91 committed:
Merge pull request #11892 from sarjon/fix-missing-category-constraint-const

Fix missing const definition

Quetzacoalt91 committed:
Merge pull request #11876 from Dan-Orchard/develop

CO: Fix lang table multi-store data issue

Quetzacoalt91 committed:
Merge pull request #11809 from olivier-monaco/misuse-of-getimagesize-1.7

CO: Fix misuse of getimagesize returned informations

Quetzacoalt91 committed:
Generate core assets

DelecroixQuentin committed:
Fix #11840

matks committed:
Fix Admin legacy tests

matks committed:
Fix unit standard legacy tests

Commit:
Merge pull request #11902 from DelecroixQuentin/develop

Fix #11840

PeeyushAgrawal committed:
issue  #11096 fixed

PeeyushAgrawal committed:
added display hook to show additional customer address fields in address block on customer addresses controller

PeeyushAgrawal committed:
Update block-address.tpl

PierreRambaud committed:
Merge pull request #11608 from Fabuloops/js-core-jquery-load-shortcut-deprecated

Fix jquery shortcut function 'load' deprecate

matks committed:
Fix CLDR tests by restoring encoding

Quetzacoalt91 committed:
Merge pull request #11865 from MathiasReker/t12

PHP unit set up tear down visibility

spolischook committed:
use env variable for admin dir with default value

no BC break

Quetzacoalt91 committed:
Merge pull request #11919 from spolischook/patch-1

use env variable for admin dir with default value

Commit:
Fix searching issue in the Credit Slips page

ansar21mallouli committed:
TE: Filters and search in catalog page

Commit:
Can't sort tax included for now

matks committed:
WIP - fix PositionControllerTest with refactoring

Commit:
Fix data provider error

Commit:
Fix lint

YosraAk committed:
TE: update autoupgrade test

Commit:
Fix undefined index

Commit:
Fix override tests by switching module order

PierreRambaud committed:
Merge pull request #11899 from matks/introduce-icon_title-for-helper-list

Introduce HelperList title_icon property

PierreRambaud committed:
Merge pull request #11907 from PeeyushAgrawalWebkul/develop

issue  #11906 Solution

Commit:
Fix test by using Module Manager

Commit:
Description has a max length and can not exceeded

Commit:
Add id_customization to a cart's cart rows association

Commit:
Add id_customization to an order's order rows association

Commit:
Add xlink reference to product_id on order_rows association

ansar21mallouli committed:
TE: filters by category in catalog page

PierreRambaud committed:
Merge pull request #11938 from eternoendless/add-ws-customization

Make customer's product customization information reachable in WebServices

tomas862 committed:
applies cs fixer

matks committed:
Fix tests by using ps_banner instead of ps_featuredproducts

PierreRambaud committed:
Merge pull request #10487 from SebBareyre/patch-34

Added breadcrumb to the manufacturers pages

PierreRambaud committed:
Update phpdoc and reuse PS_LANG_DEFAULT

matks committed:
Refactor variable names

PierreRambaud committed:
Cs fixer

MathiasReker committed:
PHP unit test case static method calls

Quetzacoalt91 committed:
Merge pull request #11985 from MathiasReker/php-unit-test-case-static-method-calls

PHP unit test case static method calls

PierreRambaud committed:
Merge pull request #11922 from khouloudbelguith/fix-issues-11921

Fix searching issue in the Credit Slips page

PierreRambaud committed:
Merge pull request #11754 from tomlev/issue-11541

Fix wrong discount details on order page / on invoice

Quetzacoalt91 committed:
Merge pull request #11936 from PierreRambaud/fix/description-max-length

Description has a max length and can not exceeded

Quetzacoalt91 committed:
Merge pull request #11932 from PierreRambaud/fix/sorting-tax-incl

Can't sort tax included for now

Quetzacoalt91 committed:
Merge pull request #11898 from matks/fix-tests-develop

Fix travis script for running tests and fix broken tests

Commit:
Remove useless parenthesis

Commit:
Fix validation

Commit:
Make Lazy Arrays clone their array access properties when being cloned

Commit:
Clone lazyarray instances before filtering them

Commit:
Add unit test to ensure filtering a LazyArray doesn't affect previous instances

matks committed:
Isolate legacy tests into suites to fix context and fixtures overlap that make tests fail

Quetzacoalt91 committed:
Merge pull request #12013 from matks/fix-unit-tests-develop-2

Isolate legacy tests into suites to fix develop branch

mwht committed:
fix typo in example nginx config file

Commit:
Merge pull request #10294 from sadlyblue/patch-14

Fix minimum order amount for secondary currency

Commit:
Merge pull request #11726 from jolelievre/dni-required

Force dni fields required if associated Country needs it

PierreRambaud committed:
Merge pull request #12015 from mwht/nginx-config-typo-bugfix

fix typo in example nginx config file

Commit:
Change thumbnail block width

mbadrani committed:
Merge pull request #12007 from YosraAk/delivery_slip

Delivery slip test

mbadrani committed:
Merge pull request #12001 from ansar21mallouli/filters_and_search_in_catalog_page

Filters in catalog page

NMAC427 committed:
Fix invalid URL encoding

Due to the wrong URL encoding the key parameter was lost during the redirect after authentication.

matks committed:
Use Symfony Cache Clear in ModuleManager instead of deprecated Cache Clearer

NMAC427 committed:
Check if the order was made by the current user

Previously anyone with the correct link could download a virtual good. Specifically:
- If no user was logged in, and either no order ID or no secure key was specified.
- If a user was logged in, but they used a download link from another user.

matks committed:
Merge pull request #10774 from tomas862/migration/international/localization/currencies_page

Migration of "International > Localization > Currencies" page listing

Quetzacoalt91 committed:
Merge pull request #10477 from mickaelandrieu/quality/custom-error-handler

Added an Error Data Listener

Quetzacoalt91 committed:
Merge pull request #12010 from eternoendless/lazyarray-clone

Make filtering of LazyArrays return a new instance

NMAC427 committed:
Fix indentation.

Commit:
Format markup

Commit:
Extract variables in scss

Commit:
Fix product title offset

Commit:
Build assets

Commit:
Slightly increase horizontal spacing between cards

Commit:
Make pre-commit execute linting only if php files have been staged

NMAC427 committed:
Conform to coding standards

PierreRambaud committed:
Merge pull request #10297 from CaptainYouz/issue-9738_product-card-image-dimension

Fix image dimensions in Product Card view

matks committed:
Merge pull request #11481 from sarjon/add-missing-category-features

Add cover & menu thumbnail images deleting for category

matks committed:
Fix CustomerFormDataProviderTest

matks committed:
Merge pull request #11723 from sarjon/migrate/customer-crud

Migrate customer Add/Edit actions

Commit:
Use single brackets for compatibility

mbadrani committed:
Merge pull request #12030 from ansar21mallouli/filter_by_category_in_catalog_page

Filters by category in catalog page

ansar21mallouli committed:
TE: Create, edit, activate restrictions country

NMAC427 committed:
Handle guest customers

Fixes a bug where guests weren't able to download virtual products.

NMAC427 committed:
Display error message

If only either the order id or the secure key are set, display an error message.

NMAC427 committed:
Fix indentation

mbadrani committed:
Merge pull request #12045 from YosraAk/update_autoupgrade

Update autoupgrade test

mbadrani committed:
Merge pull request #12049 from ansar21mallouli/branch-position

Change position, unhook and transplant a module

Quetzacoalt91 committed:
Merge pull request #12022 from matks/use-symfony-cache-clearer-for-module-manager

Use Symfony Cache Clear in ModuleManager instead of deprecated Cache …

Commit:
Fix avoid mutating a prop directlygss

Commit:
Do not fix files automatically, prevent the user from committing them until they are fixed

NMAC427 committed:
Merge if statements

PierreRambaud committed:
Merge pull request #11854 from Klemart3D/patch-1

Add customer_message table name into SQL request

matks committed:
Migrate legacy tests for Hashing

matks committed:
Use single quotes in Hashing unit test

matks committed:
Enable PHP CS Fixer on tests folder

PierreRambaud committed:
Merge pull request #12054 from matks/add-unit-tests-3

Migrate legacy unit test for Hashing

Quetzacoalt91 committed:
Merge pull request #12060 from matks/enable-cs-fixer-on-test-folder

Enable PHP CS Fixer on tests folder

Commit:
Do not show progress

Quetzacoalt91 committed:
Merge pull request #12023 from garnele007/develop

When downloading a virtual good,  check if the order was made by the current user.

PierreRambaud committed:
Build assets

Quetzacoalt91 committed:
Merge pull request #12031 from eternoendless/improve-pre-commit

Make pre-commit execute linter only if php files have been staged

Quetzacoalt91 committed:
Merge pull request #12052 from PierreRambaud/fix/serp

SERP : Avoid mutating a prop directly + fix url encode

tomlev committed:
fix static cache reset on ObjectModel

PierreRambaud committed:
Merge pull request #12079 from tomlev/test-fix

fix static cache reset on ObjectModel

tomlev commented on composer.json:
```diff
@@ -184,6 +185,9 @@
             "@composer create-test-db",
             "@php -d date.timezone=UTC ./vendor/bin/phpunit -c tests-legacy/old-controllers.xml"
         ],
+        "integration-domain-tests": [
```
this could be named "behavior-driven-tests", since we will use it for unit/integration/e2e tests

tomlev commented on tests/Integration/Domain/features/bootstrap/FeatureContext.php:
```diff
@@ -0,0 +1,114 @@
+<?php
+
```
no namespace here ?

tomlev commented on tests/Integration/Domain/behat.yml:
```diff
@@ -0,0 +1 @@
+# nothing yet here
```
what about naming this folder "behavior" or something more descriptive ?

mbadrani committed:
Merge pull request #12086 from ansar21mallouli/location_zone_country

Location: zone and country

tomlev committed:
fix conflicts

tomlev committed:
change directory name

eternoendless commented:
Replaced by #12090
